### ISSUE TITLE
Question regarding difference between checkpoints, and evaluation on megadepth

### ISSUE THREAD

**Issue by Parskatt** [facebookresearch/silk#27]
Hi, thanks for the great work. From reading your FAQ:

> `Checkpoint`. We provide two checkpoints for our VGG-4 backbone : pvgg-4.ckpt and coco-rgb-aug.ckpt (trained with different photometric augmentations). Try both and see which one performs best.

For me these produce significantly different results. On the common megadepth test split, scenes 0015 and 022 (provided by loftr authors), the coco-rgb-aug.ckpt significantly outperforms the first one. In the FAQ you mention different photmetric augmentation, is this the only difference? 

In the paper I could not find any reference to different training reginmes being used for HPatches and IMC2022, so I'm a bit confused.

**Comment by Parskatt** [comment]
I know this is a lot to ask, but could you provide reference results on mega1500 (using same ransac settings as loftr).
I'm having some difficulty understanding some of the inner workings of your codebase, and concerned that I'm getting stuff wrong.

**Comment by gleize** [comment]
Hi @Parskatt, 

> In the FAQ you mention different photmetric augmentation, is this the only difference ? In the paper I could not find any reference to different training reginmes being used for HPatches and IMC2022, so I'm a bit confused.

Yes, it's the only difference. Initially, we trained the first version (`pvgg-4.ckpt`) using similar augmentations as SuperPoint and ran our HPatches / ScanNet experiments with it. Then, during internal testing, we noticed some performance drop on outdoor images. Thus we tried a different augmentation we thought more adapted to outdoor images and got the IMC result from that.

Initially, we had a sentence mentioning that in the paper, but the mention got removed during the shortening phase. We will add it back to reduce confusion. Since we did not heavily explored / measured the effects of augmentations, we did not focus too much on it in the paper, similar to existing work (e.g. vague mention of augmentations in SuperPoint; different augmentations are not mentioned in LoFTR / R2D2, but present in code).

> On a slightly unrelated note, and I know this is a lot to ask, but could you provide reference results on mega1500 (using same ransac settings as loftr)?

I don't have enough time to set that up, however, I don't mind having a look at your code and double check how you're running it. You can either share the code in this post, or put it in a private github repository and invite me.

**Comment by Parskatt** [comment]
Thanks for the reply @gleize. I'll send you a stripped down version of what I'm doing later today, in particular I think I may have gotten the coordinate conversions/matcher settings wrong.

**Comment by Parskatt** [comment]
@gleize 
I've uploaded a stripped down version of my current implementation here: https://github.com/Parskatt/silk-mega1500-eval/tree/master

Note: I hardcode 512x512 resolution currently. This seems to work fine for the SiLK detector, I'm not sure about the descriptor.

* Results on mega1500 are *extremely* sensitive to MNN hyperparameters threshold / temperature. I'm almost wondering if the features I'm using are random? However, the weights do seem to load, and results on HPatches matches well with reported results in your paper.

**Comment by gleize** [comment]
Hi @Parskatt,

> [...] results on HPatches matches well with reported results in your paper

If that's the case, then the error likely doesn't come from your use of SiLK. I've seen you split the detection and description. It looks fine at first glance. You can always double check your keypoints / matches using our [cli tools](https://github.com/facebookresearch/silk/blob/main/doc/usage/inference.md#cli-tools) if you have a doubt.

I've also had a quick look at the benchmarking code, and there is something I don't understand. Why is the intrinsics rescaled ? (c.f. [here](https://github.com/Parskatt/silk-mega1500-eval/blob/8bc7d070a32de3c4ee50c7fdb6791c1d8a77ef74/mega1500.py#L64C10-L64C10)). In theory it shouldn't change the result because of the scale ambiguity, but I'm wondering why it's there at all, and if it could somehow affect the pose estimation in some non-trivial way (e.g. change in the inlier threshold).

Looking more in details at the [inlier threshold](https://github.com/Parskatt/silk-mega1500-eval/blob/8bc7d070a32de3c4ee50c7fdb6791c1d8a77ef74/mega1500.py#L79), it seems you are computing the mean of the focal diagonal, but also the two 0 off diagonal. This is different from LoFTR that only use the focal diagonal element (c.f. [here](https://github.com/zju3dv/LoFTR/blob/c0546af290fe87aaa75f754afe2759e0b00412ea/src/utils/metrics.py#L80)). Since most keypoint methods are sensitive to inlier threshold, this would affect your results.

> Results on mega1500 are extremely sensitive to MNN hyperparameters threshold / temperature

The temperature of 20 and a scaling factor of 1 should be optimal since it was trained with those parameters. The matching parameters tend to affect all keypoint methods, and should be tuned on a validation set. In the case of SiLK, we recommend tuning [those parameters](https://github.com/facebookresearch/silk/blob/main/doc/usage/faq.md#how-to-tune-silk-).

> Matcher uses temp 20 (different from IMC2022). If I use IMC2022 settings the results are extremely poor.

It's actually not different from IMC, it's just a bit subtle because you did split the detection and description steps. The `SCALING_FACTOR` of 1.41 should have been included in the call of this [function](https://github.com/Parskatt/silk-mega1500-eval/blob/8bc7d070a32de3c4ee50c7fdb6791c1d8a77ef74/silk_descriptor.py#L102). Since it's not included, it will default to being 1. 

Calling the model with a scaling factor of 1 and a temperature of 20 is equivalent to calling it with a scaling factor of $\sqrt{2}$ and a temperature of 10 (c.f. [here](https://github.com/facebookresearch/silk/blob/main/doc/usage/faq.md#what-is-silk_scale_factor-141-)).

> Resizing all images to 512x512

At first, I thought this would require an update of the intrinsics, but since you're operating on normalized coordinates, it should be fine.

Not conserving the aspect ratio might possibly affect the descriptors (speculation since we haven't tried it).

**Comment by Parskatt** [comment]
Thanks for the detailed review! The double descriptor scaling makes sense, didn't think about that.

Iirc all intrinsics are diagonal, i.e. no off diagonal elements. The rescaling of the intrinsics correspons to intrinsics similar to the ones in aspanformer. I remember playing around with that threshold and not finding any major differences. However, it might be the case that it matters more for silk.

As for resize, I'll try longer side resizing instead, although since you train on homographies I believe silk should be quite robust to aspect ratio. 
.
